chainer_chemistry/functions/mean_squared_error.py

```diff
@@ -0,0 +1,75 @@
+import numpy
+
+from chainer import cuda
+from chainer import function_node
+import chainer.functions
+from chainer.utils import type_check
+
+
+class MeanSquaredError(function_node.FunctionNode):
+
+    """Mean squared error (a.k.a. Euclidean loss) function."""
+
+    def __init__(self, ignore_nan=False):
+        self.task_weight = None
+        self.ignore_nan = ignore_nan
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+        type_check.expect(
+            in_types[0].dtype == numpy.float32,
+            in_types[1].dtype == numpy.float32,
+            in_types[0].shape == in_types[1].shape
+        )
+
+    def forward_cpu(self, inputs):
+        self.retain_inputs((0, 1))
+        diff = (inputs[0] - inputs[1]).ravel()
+        if self.ignore_nan:
+            diff[numpy.isnan(diff)] = 0.
+        return numpy.array(diff.dot(diff) / diff.size, dtype=diff.dtype),
+
+    def forward_gpu(self, inputs):
+        cupy = cuda.cupy
+        self.retain_inputs((0, 1))
+        diff = (inputs[0] - inputs[1]).ravel()
+        if self.ignore_nan:
+            diff[cupy.isnan(diff)] = 0.
+        return diff.dot(diff) / diff.dtype.type(diff.size),
+
+    def backward(self, indexes, gy):
+        x0, x1 = self.get_retained_inputs()
+        xp = cuda.get_array_module(x0)
+        ret = []
+        diff = x0 - x1
+        if self.ignore_nan:
+            diff = chainer.functions.where(xp.isnan(diff.array),
+                                           xp.zeros_like(diff.array), diff)
```

Comment:
I'm not sure but `xp.zeros_like(diff.array)` can be just 0, in that case we can skip creating (allocating memory of) maybe large zero array.